File: clients/client-python/gravitino/dto/requests/table_create_request.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

from dataclasses import dataclass, field
from typing import Optional

from dataclasses_json import config

from gravitino.dto.rel.column_dto import ColumnDTO
from gravitino.dto.rel.distribution_dto import DistributionDTO
from gravitino.dto.rel.indexes.index_dto import IndexDTO
from gravitino.dto.rel.indexes.json_serdes.index_serdes import IndexSerdes
from gravitino.dto.rel.json_serdes.distribution_serdes import DistributionSerDes
from gravitino.dto.rel.json_serdes.sort_order_serdes import SortOrderSerdes
from gravitino.dto.rel.partitioning.json_serdes.partitioning_serdes import (
    PartitioningSerdes,
)
from gravitino.dto.rel.partitioning.partitioning import Partitioning
from gravitino.dto.rel.sort_order_dto import SortOrderDTO
from gravitino.rest.rest_message import RESTRequest
from gravitino.utils.precondition import Precondition


@dataclass
class TableCreateRequest(RESTRequest):  # pylint: disable=too-many-instance-attributes
    """Represents a request to create a table."""

    _name: str = field(metadata=config(field_name="name"))
    _columns: list[ColumnDTO] = field(metadata=config(field_name="columns"))
    _comment: Optional[str] = field(default=None, metadata=config(field_name="comment"))
    _partitioning: Optional[list[Partitioning]] = field(
        default=None,
        metadata=config(
            field_name="partitioning",
            encoder=lambda items: [
                PartitioningSerdes.serialize(item) for item in items
            ],
            decoder=lambda values: [
                PartitioningSerdes.deserialize(value) for value in values
            ],
            exclude=lambda value: value is None,
        ),
    )
    _distribution: Optional[DistributionDTO] = field(
        default=None,
        metadata=config(
            field_name="distribution",
            encoder=DistributionSerDes.serialize,
            decoder=DistributionSerDes.deserialize,
            exclude=lambda value: value is None,
        ),
    )
    _sort_orders: Optional[list[SortOrderDTO]] = field(
        default=None,
        metadata=config(
            field_name="sortOrders",
            encoder=lambda items: [SortOrderSerdes.serialize(item) for item in items],
            decoder=lambda values: [
                SortOrderSerdes.deserialize(value) for value in values
            ],
            exclude=lambda value: value is None,
        ),
    )
    _indexes: Optional[list[IndexDTO]] = field(
        default=None,
        metadata=config(
            field_name="indexes",
            encoder=lambda items: [IndexSerdes.serialize(item) for item in items],
            decoder=lambda values: [IndexSerdes.deserialize(value) for value in values],
            exclude=lambda value: value is None,
        ),
    )
    _properties: Optional[dict[str, str]] = field(
        default=None,
        metadata=config(field_name="properties", exclude=lambda value: value is None),
    )

    def validate(self):
        Precondition.check_string_not_empty(
            self._name, '"name" field is required and cannot be empty'
        )
        if self._sort_orders:
            for sort_order in self._sort_orders:
                sort_order.validate(self._columns)

        if self._distribution:
            for expression in self._distribution.expressions():
                expression.validate(self._columns)

        if self._partitioning:
            for partitioning in self._partitioning:
                partitioning.validate(self._columns)

        for column in self._columns:
            column.validate()
        auto_increment_cols = [
            column for column in self._columns if column.auto_increment()
        ]
        auto_increment_cols_str = (
            f"[{','.join(column.name() for column in auto_increment_cols)}]"
        )
        Precondition.check_argument(
            len(auto_increment_cols) <= 1,
            "Only one column can be auto-incremented. There are multiple auto-increment "
            f"columns in your table: {auto_increment_cols_str}",
        )

        if self._indexes:
            for index in self._indexes:
                Precondition.check_argument(
                    index.type() is not None, "Index type cannot be null"
                )
                Precondition.check_argument(
                    index.field_names() is not None, "Index fieldNames cannot be null"
                )
                Precondition.check_argument(
                    len(index.field_names()) > 0,
                    "Index fieldNames length must be greater than 0",
                )

File: common/src/main/java/org/apache/gravitino/dto/requests/TableCreateRequest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.gravitino.dto.requests;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.google.common.base.Preconditions;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import javax.annotation.Nullable;
import lombok.Builder;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.ToString;
import lombok.extern.jackson.Jacksonized;
import org.apache.commons.lang3.StringUtils;
import org.apache.gravitino.dto.rel.ColumnDTO;
import org.apache.gravitino.dto.rel.DistributionDTO;
import org.apache.gravitino.dto.rel.SortOrderDTO;
import org.apache.gravitino.dto.rel.expressions.FunctionArg;
import org.apache.gravitino.dto.rel.indexes.IndexDTO;
import org.apache.gravitino.dto.rel.partitioning.Partitioning;
import org.apache.gravitino.rest.RESTRequest;

/** Represents a request to create a table. */
@Getter
@EqualsAndHashCode
@ToString
@Builder
@Jacksonized
public class TableCreateRequest implements RESTRequest {

  @JsonProperty("name")
  private final String name;

  @Nullable
  @JsonProperty("comment")
  private final String comment;

  @JsonProperty("columns")
  private final ColumnDTO[] columns;

  @Nullable
  @JsonProperty("properties")
  private final Map<String, String> properties;

  @JsonProperty("sortOrders")
  @Nullable
  private final SortOrderDTO[] sortOrders;

  @JsonProperty("distribution")
  @Nullable
  private final DistributionDTO distribution;

  @Nullable
  @JsonProperty("partitioning")
  private final Partitioning[] partitioning;

  @Nullable
  @JsonProperty("indexes")
  private final IndexDTO[] indexes;

  /** Default constructor for Jackson deserialization. */
  public TableCreateRequest() {
    this(null, null, null, null, null, null, null, null);
  }

  /**
   * Creates a new TableCreateRequest.
   *
   * @param name The name of the table.
   * @param comment The comment of the table.
   * @param columns The columns of the table.
   * @param properties The properties of the table.
   * @param sortOrders The sort orders of the table.
   * @param distribution The distribution of the table.
   * @param partitioning The partitioning of the table.
   * @param indexes The indexes of the table.
   */
  public TableCreateRequest(
      String name,
      @Nullable String comment,
      ColumnDTO[] columns,
      @Nullable Map<String, String> properties,
      @Nullable SortOrderDTO[] sortOrders,
      @Nullable DistributionDTO distribution,
      @Nullable Partitioning[] partitioning,
      @Nullable IndexDTO[] indexes) {
    this.name = name;
    this.columns = columns;
    this.comment = comment;
    this.properties = properties;
    this.sortOrders = sortOrders;
    this.distribution = distribution;
    this.partitioning = partitioning;
    this.indexes = indexes;
  }

  /**
   * Validates the {@link TableCreateRequest} request.
   *
   * @throws IllegalArgumentException If the request is invalid, this exception is thrown.
   */
  @Override
  public void validate() throws IllegalArgumentException {
    Preconditions.checkArgument(
        StringUtils.isNotBlank(name), "\"name\" field is required and cannot be empty");

    if (sortOrders != null) {
      Arrays.stream(sortOrders).forEach(sortOrder -> sortOrder.validate(columns));
    }

    if (distribution != null) {
      Arrays.stream((FunctionArg[]) distribution.expressions())
          .forEach(expression -> expression.validate(columns));
    }

    if (partitioning != null) {
      Arrays.stream(partitioning).forEach(p -> p.validate(columns));
    }

    List<ColumnDTO> autoIncrementCols =
        Arrays.stream(columns)
            .map(
                column -> {
                  column.validate();
                  return column;
                })
            .filter(ColumnDTO::autoIncrement)
            .collect(Collectors.toList());
    String autoIncrementColsStr =
        autoIncrementCols.stream().map(ColumnDTO::name).collect(Collectors.joining(",", "[", "]"));
    Preconditions.checkArgument(
        autoIncrementCols.size() <= 1,
        "Only one column can be auto-incremented. There are multiple auto-increment columns in your table: "
            + autoIncrementColsStr);

    if (indexes != null && indexes.length > 0) {
      Arrays.stream(indexes)
          .forEach(
              index -> {
                Preconditions.checkArgument(index.type() != null, "Index type cannot be null");
                Preconditions.checkArgument(
                    index.fieldNames() != null, "Index fieldNames cannot be null");
                Preconditions.checkArgument(
                    index.fieldNames().length > 0,
                    "Index fieldNames length must be greater than 0");
              });
    }
  }
}
